docs/design-docs/05-kpt-fn-exclude.md

```diff
@@ -0,0 +1,233 @@
+```
+
+```yaml
+# pipeline of Kptfile
+pipeline:
+  mutators:
+  - image: set-labels:v0.1.5
+    selectors:
+    - annotations:
+        foo: bar
+    exclude:
+    - annotations:
+        foo: baz
+```
+
+## Open Issues/Questions
+
+### How should kpt functions handle meta resources?
+
+One concern about this approach is that in order to exclude the Kptfile from being modified by horizontal
+transformations like set-namespace and set-labels, the user will have to write an exclusion field for each function.
+Arguably, the user shouldn't have to provide any special syntax for the Kptfile at all, because kpt and 
+kpt functions should be able to handle it correctly. A mitigation for this is that our own horizontal kpt functions
+can exclude the Kptfile within the function logic itself, and there will be no special handling logic in kpt. 
+
+## Alternatives Considered
+
+### A new annotation to decide if a kpt function should modify it
+
+We can introduce a new annotation that determines if functions should modify it.
+
+```yaml
+apiVersion: kpt.dev/v1
+kind: Kptfile
+metadata:
+  name: example
+  annotations:
+    config.kubernetes.io/hydration-override: false
+```
+
+If kpt sees `config.kubernetes.io/hydration-override: false`, it won't allow the functions to modify the resource. The user can add 
+this annotation to every resource that they don't want functions to modify.
+
+Pros:
+- The user would not need to write an exclusion field for each function.
+- Simple, easy to understand UX.
+
+Cons:
+- It will not be possible to control at a function level. If users want some functions to modify the Kptfile/meta resources, and others not to touch it, 
+they will have to re-organize their package.
+
+
+### Use the existing selector mechanism
+
+We can recommend that users use the existing selectors to exclude the Kptfile and functionConfigs if they don't want functions to run on them. 
+Here is an example: https://github.com/natasha41575/kpt-exclude-kptfile/blob/main/Kptfile. The Kptfile looks like:
+
+```yaml
+apiVersion: kpt.dev/v1
+kind: Kptfile
+metadata:
+  name: example
+pipeline:
+  mutators:
+  - image: gcr.io/kpt-fn/set-labels:unstable
+    configPath: functionconfig.yaml
+    selectors:
+    - apiVersion: "batch/v1"
+    - apiVersion: "apps/v1"
+    - apiVersion: "v1"
+    - apiVersion: "example.com/v1"
+```
+
+Some problems with this approach:
+- It is verbose.
+- Verbosity increases if the user has more resources in the package.
+- The user will have to keep track of an exhaustive list of all their resources in their package just to exclude the Kptfile.
+- kpt and kpt functions should already know about the Kptfile, so this overhead on the user's side is unnecessary.
+
+### Rely on the existing annotation config.kubernetes.io/local-config
+
+We can exclude all local-config resources (i.e. resources with the annotation `config.kubernetes.io/local-config: true`) from functions. This annotation is 
```

Comment:
This would break all functions that use local config pseudo-resources as implicit inputs. I had to workaround this behavior when we built the porch prototype.

Reply:
My bad. I proposed this to simplify the annotations users need to deal with.